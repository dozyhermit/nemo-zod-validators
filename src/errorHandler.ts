import { z } from 'zod';
import type { ZodError } from './types';

export const DEFAULT_ERROR_MESSAGE = 'Validation has failed';

// TECHNICAL DEBT: remove .flatten(), use z.flattenError instead
const getErrors = (
  error: ZodError['error'],
  key: keyof z.core.$ZodFlattenedError<string, unknown>
) => {
  if (typeof error?.flatten !== 'function') {
    return undefined;
  }

  if (Object.values(error?.flatten()?.[key] || {}).length <= 0) {
    return undefined;
  }

  return error.flatten()[key];
};

export const errorHandlerWithSchema = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'fieldErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithSchemaFormErrors = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'formErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithTransform = () =>
  Response.json(
    { message: 'Invalid request', success: false },
    {
      status: 400,
    }
  );
