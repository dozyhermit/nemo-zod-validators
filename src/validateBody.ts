import { MiddlewareFunctionProps } from '@rescale/nemo';
import type { ValidateWithSchema } from './types';
import { validateGeneric } from './validateGeneric';

type ValidateBody<T> = ValidateWithSchema<T>;

/**
 * Validates the `request.body` object inside a `NextRequest` typed request using `zod.safeParse`.
 *
 * @param {T} schema The schema doing the validating
 * @param {Function} errorHandler A custom error function
 */
export const validateBody = <T>({ schema, errorHandler }: ValidateBody<T>) => {
  return async ({ request }: MiddlewareFunctionProps) =>
    validateGeneric({ data: request.body as T, schema, errorHandler });
};
