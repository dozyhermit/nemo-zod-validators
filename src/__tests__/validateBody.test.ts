import { validateBody } from '../validateBody';
import { errorHandler } from './mocks/errorHandler';
import {
  type SchemaType,
  schema,
  schemaWithIntentionallyBrokenSafeParse,
} from './mocks/schema';

describe('validateBody', () => {
  test('passes validation with correct data', async () => {
    const action = validateBody<SchemaType>({ schema });

    const result = await action({
      request: {
        // @ts-ignore type inconsistencies due to mocking body
        body: { Number: 123, String: 'Hello', AdvancedString: '123456' },
      },
    });

    expect(result.status).toBe(200);
  });

  test('triggers validation error with incorrect data', async () => {
    const action = validateBody<SchemaType>({ schema });

    const result = await action({
      request: {
        // @ts-ignore type inconsistencies due to mocking body
        body: { Number: 123, String: 'Hello', AdvancedString: 'World' },
      },
    });

    expect(await result.json()).toMatchSnapshot();
  });

  test('triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE', async () => {
    const action = validateBody<SchemaType>({
      // @ts-ignore
      schema: schemaWithIntentionallyBrokenSafeParse,
    });

    const result = await action({
      request: {
        // @ts-ignore type inconsistencies due to mocking body
        body: { Number: 123, String: 'Hello', AdvancedString: 'World' },
      },
    });

    expect(await result.json()).toMatchSnapshot();
  });

  describe('with custom error handler', () => {
    test('passes validation with correct data', async () => {
      const action = validateBody<SchemaType>({ schema, errorHandler });

      const result = await action({
        request: {
          // @ts-ignore type inconsistencies due to mocking body
          body: { Number: 123, String: 'Hello', AdvancedString: '123456' },
        },
      });

      expect(result.status).toBe(200);
    });

    test('triggers validation error with incorrect data', async () => {
      const action = validateBody<SchemaType>({ schema, errorHandler });

      const result = await action({
        request: {
          // @ts-ignore type inconsistencies due to mocking body
          body: { Number: 123, String: 'Hello', AdvancedString: 'World' },
        },
      });

      expect(await result.json()).toMatchSnapshot();
    });
  });
});
