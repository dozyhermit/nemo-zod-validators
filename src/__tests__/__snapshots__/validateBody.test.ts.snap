// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`validateBody triggers validation error with incorrect data 1`] = `
{
  "message": {
    "AdvancedString": [
      "Invalid AdvancedString regex",
    ],
  },
}
`;

exports[`validateBody triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE 1`] = `
{
  "message": "Validation has failed",
}
`;

exports[`validateBody with custom error handler triggers validation error with incorrect data 1`] = `
{
  "message": "Invalid validation!",
}
`;
